### PR TITLE
Mesh motion rework

### PR DESCRIPTION
**Pull request by ashesh2512**
This pull request replaces the existing mesh motion framework with a more generic and reference frame-based mesh motion framework wherein any motion frame is classified as inertial or non-inertial. Any motion in 3D can be represented using a 4x4 transformation matrix. The new framework makes use of such matrices to build composite transformations useful for wind farm simulations.

1) The wall boundary condition, in presence of mesh motion, for the Momentum equation system has been completely replaced to move away from having to define a wind energy auxiliary function that would recompute the fluid velocity at the wall. Instead, the mesh velocity is directly copied over to the fluid velocity at the wall. 

2) Gold files and input files for regression tests involving mesh motion, specifically tests 15, 16, 17, 18, 69, and 78 have been updated. Furthermore, certain regression test files had left over `.remote` files from a possible accidental commit. Those have been removed.
-- For tests 15 (dgNonConformalElemCylinder) and 78 (uqSlidingMeshDG), the differences with the new mesh motion framework are negligible and can be accorded to round off errors due to change in the mesh motion algorithm. 
-- For tests 18 (dgNonConformalThreeBlade) and 69 (dgncThreeBladeHypre) the differences are due to an existing bug highlighted in #70. This pull request fixes that bug.
-- For tests 16 (dgNonConformalFluids) and 17(dgNonConformalFluidsEdge) the differences are a result of a bug in the respective input files wherein the wind_energy option was never specified for the wall boundary conditions. 

3) Unit tests have been added for the different motions (rotation, scaling, translation) introduced in this pull request. 

4) The documentation has been updated to include an example of mesh motion definition in an input file in line with the new mesh motion framework. Below is a preview of the documentation changes
https://ashesh-nalu-fork.readthedocs.io/en/mesh_motion_rework/source/user/nalu_run/nalu_inp.html#solution-options

### CODE REVIEW

File: include/SolutionOptions.h [hunk 1]
```diff
@@ -204,7 +203,7 @@ class SolutionOptions
   std::map<std::string, std::string> inputVarFromFileMap_;
 
   // mesh motion
-  std::map<std::string, MeshMotionInfo *> meshMotionInfoMap_;
+  YAML::Node meshMotionNode_;
```
[the remark this answers is not in the document]
Comment (ashesh2512) [reply include/SolutionOptions.h]: @sayerhs I need the realm to initialize `meshMotionAlg_` and other mesh motion related classes therein. `SolutionOptions` unfortunately does not own an instance of `Realm`, so I am not sure what would be the best way to go about this. One option I can think of -- Create  `meshMotionAlg_` in `SolutionOptions` but then set the relevant member variables (meta_data, bulk_data) in `Realm::load()`. Would that be cleaner ?

File: src/LowMachEquationSystem.C [hunk 2]
```diff
@@ -1739,55 +1738,70 @@ MomentumEquationSystem::register_wall_bc(
   VectorFieldType *theBcField = &(meta_data.declare_field<VectorFieldType>(stk::topology::NODE_RANK, bcFieldName));
   stk::mesh::put_field_on_mesh(*theBcField, *part, nDim, nullptr);
 
-  // extract the value for user specified velocity and save off the AuxFunction
-  AuxFunction *theAuxFunc = NULL;
-  std::string velocityName = "velocity";
+  if(realm_.solutionOptions_->meshMotion_) {
```
[the remark this answers is not in the document]
Comment (ashesh2512) [reply src/LowMachEquationSystem.C]: @sayerhs So currently in the code if there is a wall boundary condition in the input file,`wall_user_data` is a must. The definition of `wall_user_data` in the input file can be empty but there always needs to be a `wall_user_data` key word. So for a user to move from an auxiliary function based definition of `velocity` at the wall to using the `mesh_velocity` they would need to remember to remove the key word "velocity" from the `wall_user_data` in the input file but not `wall_user_data` itself. This seems very prone to user-induced errors. What do you think ? I guess it comes down to the question you asked, why and when would one not want a no-slip condition along the moving interfaces ?

File: src/mesh_motion/MotionBase.C [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  = {{{1,0,0,0},
+      {0,1,0,0},
+      {0,0,1,0},
+      {0,0,0,1}}};
+
+MotionBase::transMatType MotionBase::add_motion(
+    const transMatType& motionL,
+    const transMatType& motionR)
+{
+  transMatType comp_trans_mat_ = {};
```
[the remark this answers is not in the document]
Comment (ashesh2512) [reply src/mesh_motion/MotionBase.C]: I believe `=` with initializer lists results in a copy construction.

File: src/LowMachEquationSystem.C (hunk 2 above)
[the remark this answers is not in the document]
Comment (sayerhs) [reply src/LowMachEquationSystem.C]: So two issues I notice: 1. how will this formulation affect when the user wants to use wall function, and 2. how will this affect where there are moving and non-moving walls? Terrain with rotating turbines comes to mind. Are we taking care of registering `mesh_velocity` field on all parts even if they are not registered through mesh_motion section so that the non-moving walls essentially have a no-op?

File: src/mesh_motion/MotionBase.C (hunk 3 above)
[the remark this answers is not in the document]
Comment (sayerhs) [reply src/mesh_motion/MotionBase.C]: Sorry, I entered the comment at the wrong place. I was wondering about the `return` statement, will that result in a move or a copy again?

[the remark this answers is not in the document]
Comment (ashesh2512) [reply src/mesh_motion/MotionBase.C]: My understanding has always been that compilers will perform return value optimization when I return a local object by value.

File: src/LowMachEquationSystem.C (hunk 2 above)
[the remark this answers is not in the document]
Comment (ashesh2512) [reply src/LowMachEquationSystem.C]: @sayerhs 

1) This formulation currently prevents the user from using a wall function for the momentum equation if mesh motion has been specified. I see that being undesirable.  How about we let the user specify `velocity: mesh_motion` under `wall_user_data` ? Choosing this option would just copy the mesh velocity over to the fluid velocity. While this does introduce an additional variable for the user to set, it would allow the user to choose how the velocity at the wall is handled. If you agree with this approach, please let me know and I can make the appropriate changes. 

2) The mesh motion algorithm does not handle registration of `mesh_velocity`. In the scenario that there is mesh motion, Realm::register_nodal_fields() registers `mesh_velocity` for all part names returned by Realm::get_physics_target_names().

File: include/mesh_motion/MotionBase.h [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  /** Computed centroid
+   *
+   * A 3x1 vector storing the centroid computed respective
+   * to a collection of parts defined in the input file
+   */
+  threeDVecType origin_ = {{0.0,0.0,0.0}};
+
+  double startTime_{0.0};
+  double endTime_{DBL_MAX};
+  const double eps_{1e-14};
```
[the remark this answers is not in the document]
Comment (ashesh2512) [reply include/mesh_motion/MotionBase.h]: Thanks for pointing this out. This has been resolved in https://github.com/Exawind/nalu-wind/pull/77/commits/3c9cb9775ad9c549d8e2409d7d39dd23d0026ef2

File: src/mesh_motion/MeshMotionAlg.C [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+      if( it ==  frameNames.end() )
+        throw std::runtime_error("MeshMotion: Invalid reference frame: " + refFrameName);
+
+      refFrameMap_[i] = std::distance(frameNames.begin(), it);
+    }
+  }
+}
+
+void MeshMotionAlg::initialize( const double time )
```
[the remark this answers is not in the document]
Comment (ashesh2512) [reply src/mesh_motion/MeshMotionAlg.C]: I agree. I have fixed this in https://github.com/Exawind/nalu-wind/pull/77/commits/94172607cd21c3fae55ffe6e0168d30212fd1071

File: src/Realm.C [hunk 6]
```diff
@@ -285,7 +287,7 @@ Realm::~Realm()
   }
 
   // any bc data
-  std::vector<AuxFunctionAlgorithm *>::iterator iaux;
+  std::vector<Algorithm *>::iterator iaux;
```
[the remark this answers is not in the document]
Comment (ashesh2512) [reply src/Realm.C]: This has been resolved in https://github.com/Exawind/nalu-wind/pull/77/commits/792682d447854de78d8ab73fd63628ee73ae57dd

File: src/Realm.C [hunk 7]
```diff
@@ -323,6 +325,8 @@ Realm::~Realm()
 
   if (nullptr != bdyLayerStats_) delete bdyLayerStats_;
 
+  if ( nullptr != meshMotionAlg_ ) delete meshMotionAlg_;
```
[the remark this answers is not in the document]
Comment (ashesh2512) [reply src/Realm.C]: I didn't see a need for `shared_ptr`. I have changed `meshMotionAlg_` to be a `unique_ptr` in https://github.com/Exawind/nalu-wind/pull/77/commits/792682d447854de78d8ab73fd63628ee73ae57dd

File: include/mesh_motion/MeshMotionAlg.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+   *  Size is the number of motion groups in input file
+   */
+  std::vector<std::unique_ptr<FrameBase>> frameVec_;
+
+  /** Reference frame map
+   *
+   *  Map between frame indices and corresponding reference frame indices
+   *  Size is the number of motion groups with reference frames
+   */
+  std::map<int, int> refFrameMap_;
```
[the remark this answers is not in the document]
Comment (ashesh2512) [reply include/mesh_motion/MeshMotionAlg.h]: Thanks for pointing this out. I have updated the code in https://github.com/Exawind/nalu-wind/pull/77/commits/94172607cd21c3fae55ffe6e0168d30212fd1071

File: include/SolutionOptions.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (sayerhs) [reply include/SolutionOptions.h]: How about making `mesh_motion` section in input file a part of `realm:` like `actuator:`, `abl_forcing:`, etc.? Since Realm owns the `meshMotionAlg_` it would remove SolutionOptions from being involved at all.

File: src/LowMachEquationSystem.C (hunk 2 above)
[the remark this answers is not in the document]
Comment (sayerhs) [reply src/LowMachEquationSystem.C]: OK. Let's leave the code as you have it right now and revisit if we see a use-case where we would want a different behavior.

File: include/SolutionOptions.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashesh2512) [reply include/SolutionOptions.h]: I have opened an issue for this #78. Will start working on it soon.

[the remark this answers is not in the document]
Comment (ashesh2512) [reply include/SolutionOptions.h]: This issue has been resolved in https://github.com/Exawind/nalu-wind/pull/77/commits/43f9b07074f7b3a868e2aea6b03b4bf75e1f1654 and https://github.com/Exawind/nalu-wind/pull/77/commits/d168accf15240cf1d7275bb1a3f4c10f6ca1b868

File: include/mesh_motion/FrameInertial.h [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    FrameInertial(const FrameInertial&) = delete;
+
+    void compute_transformation(const double);
+
+    /** Inertial frame
+     *
+     * A 4x4 matrix that defines the composite inertial frame
+     * It is initialized to an identity matrix
+     */
+    MotionBase::transMatType inertialFrame_ = MotionBase::identityMat_;
```
[the remark this answers is not in the document]
Comment (ashesh2512) [reply include/mesh_motion/FrameInertial.h]: Thanks for catching this. This has been resolved in https://github.com/Exawind/nalu-wind/pull/77/commits/06144bdc2282584dfbda0b842bb83f2724aa70e2

File: src/mesh_motion/FrameBase.C [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+      throw std::runtime_error(
+        "FrameBase: Invalid part name encountered: " + partNamesVec_[i]);
+    else
+      partVec_[i] = part;
+  }
+
+  // compute and set centroid if requested
+  if(computeCentroid_) {
+    std::vector<double> computedCentroid(3,0.0);
+    realm_.compute_centroid_on_parts( partNamesVec_, computedCentroid );
```
[the remark this answers is not in the document]
Comment (ashesh2512) [reply src/mesh_motion/FrameBase.C]: Thanks for the suggestion. I have performed the appropriate changes in https://github.com/Exawind/nalu-wind/pull/77/commits/96eae73d8373d4afd2f96c9aa7b2e2f86f0a34e9
